astar_path_planner/src/astar_path_planner.cpp

```diff
@@ -0,0 +1,186 @@
+// Copyright 2021 RoboJackets
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+// THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+#include "astar_path_planner/astar_path_planner.hpp"
+#include <memory>
+#include <set>
+#include <vector>
+#include <queue>
+
+namespace astar_path_planner
+{
+
+void AStarPathPlanner::DeclareParameters(rclcpp_lifecycle::LifecycleNode::SharedPtr node)
+{
+  node->declare_parameter("goal_threshold", 0.05);
+  node->declare_parameter("grid_size", 0.01);
+}
+
+AStarPathPlanner::AStarPathPlanner(
+  rclcpp_lifecycle::LifecycleNode::SharedPtr node,
+  std::shared_ptr<nav2_costmap_2d::Costmap2DROS> ros_costmap)
+: logger_(node->get_logger()),
+  ros_costmap_(ros_costmap)
+{
+  goal_threshold_ = node->get_parameter("goal_threshold").as_double();
+  grid_size_ = node->get_parameter("grid_size").as_double();
+}
+
+std::vector<Point> AStarPathPlanner::Plan(const Point & start, const Point & goal)
+{
+  std::set<Point, PointComparator> expanded{PointComparator{}};
+  std::priority_queue<FrontierEntry, std::vector<FrontierEntry>,
+    FrontierEntryComparator> frontier{FrontierEntryComparator{}};
+
+  goal_ = goal;
+
+  std::vector<Point> initial_path = {start};
+  frontier.push({initial_path, GetPathCost(initial_path)});
+
+  while (!frontier.empty()) {
+    const auto [path, cost] = frontier.top();
+    frontier.pop();
+    const auto last_state = path.back();
+
+    if (expanded.find(last_state) == expanded.end()) {
+      expanded.insert(last_state);
+
+      if (IsGoal(last_state)) {
+        std::vector<Point> reduced_path{path};
+        ReducePath(reduced_path);
+        return reduced_path;
```

Comment:
I think we should return both the reduced path and the raw path for visualization purposes.

Reply:
hhmmmmmmmm.... that's probably a good idea.
My instinct then is to separate the reduce path function out of this class entirely and the next layer up can explicitly call it and have access to both.